### PR TITLE
Flutter channels for iOS

### PR DESCRIPTION
**Pull request by mravn-google**
Implementation of the FlutterChannel API with pluggable codecs for iOS.

Ready for design review. Tests, better documentation, and removal of old API to come.

There is a companion PR for flutter/flutter: https://github.com/flutter/flutter/pull/8695

### CODE REVIEW

File: shell/platform/darwin/common/string_conversions.mm [hunk 1]
```diff
@@ -6,20 +6,33 @@
 
 namespace shell {
 
-std::vector<uint8_t> GetVectorFromNSString(NSString* string) {
-  if (!string.length)
+NSData* GetNSDataFromNSString(NSString* string) {
+  if (!string)
```
Comment (chinmaygarde) [inline shell/platform/darwin/common/string_conversions.mm]: This can't be right. You are just checking if string is not `nil`. All valid strings will give you `nil` `NSData` instances.

Once you do fix the logic, you can write this as `[string length] == 0`. If `string` is `nil`, you will end up sending a message to `nil` which is fine. If it is not `nil`, you will do the length check.

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        else
+          reply([_codec encodeSuccessEnvelope:result]);
+      });
+    }
+  ];
+}
+
+- (void) handleStreamWith:(FlutterStreamHandler)handler {
+  [_controller
+    handleBinaryMessagesOnChannel:_name
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: Please make sure to run `clang-format` on the patch before you commit.

File: shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FLUTTERCODECS_H_
+#define FLUTTER_FLUTTERCODECS_H_
+
+#import <Foundation/Foundation.h>
+#include "FlutterMacros.h"
+
+FLUTTER_EXPORT
+@protocol FlutterMessageCodec
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h]: Please use the standard `NSCoder` interface along with the `NSCoding` protocol. Inventing our own pattern is fine but the use becomes immediately obvious to anyone used to `Foundation` APIs.

So the codecs defined below would be `NSCoder` subclasses and the method calls, errors, etc. would `NSCoding` compliant.

File: shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+  FlutterStandardWriter* writer = [FlutterStandardWriter withData:data];
+  [writer writeValue:message];
+  return data;
+}
+
+- (id) decode:(NSData*)message {
+  FlutterStandardReader* reader = [FlutterStandardReader withData:message];
+  id value = [reader readValue];
+  if ([reader hasMore]) {
+    @throw [NSException exceptionWithName:NSInvalidArgumentException
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm]: The Flutter Objective-C framework is exception safe and I would like to keep it that way. Can we update this (and other methods that may throw exceptions) with another argument that takes an error please. Something like `error:(NSError **)error` that returns the error and returns nil here would work fine (with the corresponding error checking of course).

File: shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+  [writer writeValue:error.details];
+  return data;
+}
+
+- (FlutterMethodCall*) decodeMethodCall:(NSData*)message {
+  FlutterStandardReader* reader = [FlutterStandardReader withData:message];
+  id value1 = [reader readValue];
+  id value2 = [reader readValue];
+  if ([reader hasMore]) {
+    @throw [NSException exceptionWithName:NSInvalidArgumentException
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm]: Same comment about the exceptions.

File: shell/platform/darwin/ios/framework/Source/FlutterStandardCodec_Internal.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef SHELL_PLATFORM_IOS_FRAMEWORK_SOURCE_FLUTTERSTANDARDCODECINTERNAL_H_
+#define SHELL_PLATFORM_IOS_FRAMEWORK_SOURCE_FLUTTERSTANDARDCODECINTERNAL_H_
+
+#include "flutter/shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h"
+
+namespace shell {
+  const UInt8 kNIL = 0;
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterStandardCodec_Internal.h]: Please use a typed enum here.

**Comment by Hixie** [comment]
You will need to update the license golden file with your new file names. There's a diff in the Travis logs that you can apply to your tree to do that.

**Comment by chinmaygarde** [comment]
Also, please add some unit tests. An executable target that depends on `//flutter/testing` should help you get started.

File: shell/platform/darwin/common/string_conversions.mm (hunk 1 above)
Comment (mravn-google) [inline shell/platform/darwin/common/string_conversions.mm]: Done

File: shell/platform/darwin/common/string_conversions.mm [hunk 7]
```diff
@@ -6,20 +6,33 @@
...
+      return nil;
+  NSData* data = [[NSData alloc] initWithBytes:string.UTF8String length:string.length];
```
Comment (mravn-google) [inline shell/platform/darwin/common/string_conversions.mm]: Done

File: shell/platform/darwin/common/string_conversions.mm [hunk 8]
```diff
@@ -6,20 +6,33 @@
...
-  return std::vector<uint8_t>(bytes, bytes + strlen(chars));
+  const uint8_t* bytes = reinterpret_cast<const uint8_t*>(data.bytes);
+  return std::vector<uint8_t>(bytes, bytes + data.length);
 }
 
-NSString* GetNSStringFromVector(const std::vector<uint8_t>& buffer) {
-  NSString* string = [[NSString alloc] initWithBytes:buffer.data()
-                                              length:buffer.size()
+NSString* GetNSStringFromNSData(NSData* data) {
+  NSString* string = [[NSString alloc] initWithBytes:data.bytes
```
Comment (mravn-google) [inline shell/platform/darwin/common/string_conversions.mm]: Done.

File: shell/platform/darwin/common/string_conversions.mm [hunk 9]
```diff
@@ -6,20 +6,33 @@
...
+                                              length:data.length
                                             encoding:NSUTF8StringEncoding];
   [string autorelease];
   return string;
 }
 
+NSData* GetNSDataFromVector(const std::vector<uint8_t>& buffer) {
+  NSData* data = [[NSData alloc] initWithBytes:buffer.data()
+                                        length:buffer.size()];
+  [data autorelease];
```
Comment (mravn-google) [inline shell/platform/darwin/common/string_conversions.mm]: Done.

File: shell/platform/darwin/ios/framework/Headers/FlutterChannels.h [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "FlutterCodecs.h"
+#include "FlutterViewController.h"
+
+typedef void (^FlutterReplyHandler)(id reply);
+typedef void (^FlutterMessageHandler)(id message, FlutterReplyHandler replyHandler);
+
+FLUTTER_EXPORT
+@interface FlutterMessageChannel : NSObject
++ (instancetype) withController:(FlutterViewController*)controller
```
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Headers/FlutterChannels.h]: Done.

File: shell/platform/darwin/ios/framework/Headers/FlutterChannels.h [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+                           name:(NSString*)name
+                          codec:(NSObject<FlutterMessageCodec>*)codec;
+- (void) send:(id)message andHandleReplyWith:(FlutterReplyHandler)handler;
```
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Headers/FlutterChannels.h]: Done.

File: shell/platform/darwin/ios/framework/Headers/FlutterChannels.h [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+- (void) handleMessagesWith:(FlutterMessageHandler)handler;
```
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Headers/FlutterChannels.h]: Done.

File: shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h (hunk 3 above)
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h]: I'll look into it. I first dismissed NDCoder/NSCoding as too big a gun for this simple situation. Please be aware that plugin writers are not exposed to the internals of the Flutter standard codec, except they are given a list of supported values as part of the documentation.

File: shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+@interface FlutterJSONMessageCodec : NSObject<FlutterMessageCodec>
+@end
+
+FLUTTER_EXPORT
+@interface FlutterStandardMessageCodec : NSObject<FlutterMessageCodec>
+@end
+
+FLUTTER_EXPORT
+@interface FlutterMethodCall : NSObject
++ (instancetype) withMethod:(NSString*)method andArguments:(id)arguments;
```
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h]: Done.

File: shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+@property(readonly) NSString* method;
+@property(readonly) id arguments;
+@end
+
+FLUTTER_EXPORT
+@interface FlutterError : NSObject
++ (instancetype) withCode:(NSString*)code message:(NSString*)message details:(id)details;
```
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h]: Done.

File: shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+@property(readonly) NSString* code;
+@property(readonly) NSString* message;
+@property(readonly) id details;
+@end
+
+FLUTTER_EXPORT
+@interface FlutterStandardTypedData : NSObject
++ (instancetype) withBytes:(NSData*)data;
```
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h]: Done.

File: shell/platform/darwin/ios/framework/Headers/FlutterMessageChannel.h [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+#include "FlutterMacros.h"
+#include "FlutterMessageCodec.h"
+#include "FlutterViewController.h"
+
+typedef void (^FlutterReplyHandler)(id reply);
+typedef void (^FlutterMessageHandler)(id message, FlutterReplyHandler replyHandler);
+
+FLUTTER_EXPORT
+@interface FlutterMessageChannel : NSObject
++ (instancetype) withController:(FlutterViewController*)controller
```
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Headers/FlutterMessageChannel.h]: Done.

File: shell/platform/darwin/ios/framework/Headers/FlutterViewController.h [hunk 17]
```diff
@@ -31,6 +32,13 @@ FLUTTER_EXPORT
 
 - (void)removeMessageListener:(NSObject<FlutterMessageListener>*)listener;
 
+- (void)sendBinaryMessage:(NSData*)message
+                onChannel:(NSString*)channelName
+       andHandleReplyWith:(FlutterBinaryReplyHandler)handler;
```
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Headers/FlutterViewController.h]: Done.

File: shell/platform/darwin/ios/framework/Headers/FlutterViewController.h [hunk 18]
```diff
@@ -31,6 +32,13 @@ FLUTTER_EXPORT
...
+
+- (void)handleBinaryMessagesOnChannel:(NSString*)channelName
+                          withHandler:(FlutterBinaryMessageHandler)handler;
```
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Headers/FlutterViewController.h]: Done.

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm (hunk 2 above)
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: Ah, nice! Done.

File: shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm (hunk 4 above)
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm]: Pending.

File: shell/platform/darwin/ios/framework/Source/FlutterStandardCodec_Internal.h (hunk 6 above)
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Source/FlutterStandardCodec_Internal.h]: Done.

File: shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm (hunk 5 above)
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm]: Pending.

File: shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm (hunk 4 above)
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm]: These error situations represent programming errors. Do we really want to handle those using `NSError`s?

Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm]: In that case, we can use an `NSAssert` or an `FTL_CHECK`.

File: shell/platform/darwin/common/buffer_conversions.h [hunk 19]
```diff
@@ -11,10 +11,10 @@
 
 namespace shell {
 
-NSData* GetNSDataFromNSString(NSString* string);
+// NSData* GetNSDataFromNSString(NSString* string);
```
Comment (chinmaygarde) [inline shell/platform/darwin/common/buffer_conversions.h]: I don't think you meant to leave these comments in here.

File: shell/platform/darwin/ios/framework/Headers/FlutterChannels.h [hunk 20]
```diff
@@ -5,19 +5,18 @@
...
 typedef void (^FlutterMessageHandler)(id message,
                                       FlutterReplyHandler replyHandler);
 
 FLUTTER_EXPORT
 @interface FlutterMessageChannel : NSObject
-+ (instancetype)messageChannelWithController:(FlutterViewController*)controller
-                                        name:(NSString*)name
-                                       codec:(NSObject<FlutterMessageCodec>*)
-                                                 codec;
++ (instancetype)messageChannelNamed:(NSString*)name
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Headers/FlutterChannels.h]: Great. These names are so much better!

File: shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h (hunk 3 above)
Comment (mravn-google) [inline shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h]: I have not done this yet. Question is if it makes sense to postpone that refactoring to unblock people waiting for platform/flutter channels.

Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h]: Your call. I think we can do this in a separate patch.
